# Porblem 1

In [1]:
import time

## Run the Script as it is

In [2]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


19.77926993370056
Epoch: 1/5.. Training loss: 0.488.. Test loss: 0.423.. Test Accuracy: 0.840
19.79761290550232
Epoch: 2/5.. Training loss: 0.373.. Test loss: 0.396.. Test Accuracy: 0.854
20.1186466217041
Epoch: 3/5.. Training loss: 0.339.. Test loss: 0.373.. Test Accuracy: 0.866
20.131465196609497
Epoch: 4/5.. Training loss: 0.316.. Test loss: 0.361.. Test Accuracy: 0.872
20.082324266433716
Epoch: 5/5.. Training loss: 0.297.. Test loss: 0.371.. Test Accuracy: 0.867
118.79080986976624


## Batch Size 16

In [3]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 16, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 16, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


27.752448081970215
Epoch: 1/5.. Training loss: 0.476.. Test loss: 0.448.. Test Accuracy: 0.837
28.701561212539673
Epoch: 2/5.. Training loss: 0.375.. Test loss: 0.404.. Test Accuracy: 0.855
29.306257724761963
Epoch: 3/5.. Training loss: 0.342.. Test loss: 0.390.. Test Accuracy: 0.860
28.63768243789673
Epoch: 4/5.. Training loss: 0.317.. Test loss: 0.367.. Test Accuracy: 0.868
28.957837343215942
Epoch: 5/5.. Training loss: 0.303.. Test loss: 0.361.. Test Accuracy: 0.874
163.80136346817017


## Batch Size 64

In [4]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 64, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 64, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


16.528880834579468
Epoch: 1/5.. Training loss: 0.500.. Test loss: 0.440.. Test Accuracy: 0.843
16.736101388931274
Epoch: 2/5.. Training loss: 0.382.. Test loss: 0.427.. Test Accuracy: 0.849
16.673275470733643
Epoch: 3/5.. Training loss: 0.347.. Test loss: 0.371.. Test Accuracy: 0.863
16.656588792800903
Epoch: 4/5.. Training loss: 0.319.. Test loss: 0.355.. Test Accuracy: 0.873
16.606858253479004
Epoch: 5/5.. Training loss: 0.300.. Test loss: 0.363.. Test Accuracy: 0.867
97.05978083610535


## Learning Rate 0.01

In [5]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.01)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


20.39034152030945
Epoch: 1/5.. Training loss: 0.583.. Test loss: 0.535.. Test Accuracy: 0.817
20.72736096382141
Epoch: 2/5.. Training loss: 0.503.. Test loss: 0.512.. Test Accuracy: 0.826
20.818732261657715
Epoch: 3/5.. Training loss: 0.486.. Test loss: 0.509.. Test Accuracy: 0.835
20.53863549232483
Epoch: 4/5.. Training loss: 0.473.. Test loss: 0.548.. Test Accuracy: 0.823
20.832820177078247
Epoch: 5/5.. Training loss: 0.456.. Test loss: 0.510.. Test Accuracy: 0.834
119.51724076271057


## Learning Rate 0.0001

In [6]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.Adam(model.parameters(), lr = 0.0001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


20.479434728622437
Epoch: 1/5.. Training loss: 0.651.. Test loss: 0.514.. Test Accuracy: 0.815
20.618394374847412
Epoch: 2/5.. Training loss: 0.455.. Test loss: 0.460.. Test Accuracy: 0.835
20.535950422286987
Epoch: 3/5.. Training loss: 0.416.. Test loss: 0.441.. Test Accuracy: 0.841
20.329606533050537
Epoch: 4/5.. Training loss: 0.392.. Test loss: 0.419.. Test Accuracy: 0.848
20.27568006515503
Epoch: 5/5.. Training loss: 0.374.. Test loss: 0.409.. Test Accuracy: 0.853
118.01114583015442


## Optimizer SGD

In [7]:
time1 = time.time()
#import the nescessary libs
import numpy as np
import torch

# Loading the Fashion-MNIST dataset
from torchvision import datasets, transforms

# Get GPU Device
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")


# Define a transform to normalize the data
transform = transforms.Compose([transforms.ToTensor(),
                                    transforms.Normalize((0.5,), (0.5,))
                                                                   ])
# Download and load the training data
trainset = datasets.FashionMNIST('MNIST_data/', download = True, train = True, transform = transform)
testset = datasets.FashionMNIST('MNIST_data/', download = True, train = False, transform = transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)

# Examine a sample
dataiter = iter(trainloader)
images, labels = dataiter.next()

# Define the network architecture
from torch import nn, optim
import torch.nn.functional as F

model = nn.Sequential(nn.Linear(784, 128),
                      nn.ReLU(),
                      nn.Linear(128, 10),
                      #nn.LogSoftmax(dim = 1)
                     )
model.to(device)

# Define the loss
criterion = nn.CrossEntropyLoss()

# Define the optimizer
optimizer = optim.SGD(model.parameters(), lr = 0.001)

# Define the epochs
epochs = 5

train_losses, test_losses = [], []

for e in range(epochs):
  running_loss = 0
  train_time = time.time()
  for images, labels in trainloader:
    # Flatten Fashion-MNIST images into a 784 long vector
    images = images.to(device)
    labels = labels.to(device)
    images = images.view(images.shape[0], -1)
    
    # Training pass
    optimizer.zero_grad()
    
    output = model.forward(images)
    loss = criterion(output, labels)
    loss.backward()
    optimizer.step()
    
    running_loss += loss.item()
  else:
    train_time2 = time.time()
    print(str(train_time2 - train_time))
    test_loss = 0
    accuracy = 0
    
    # Turn off gradients for validation, saves memory and computation
    with torch.no_grad():
      # Set the model to evaluation mode
      model.eval()
      
      # Validation pass
      for images, labels in testloader:
        images = images.to(device)
        labels = labels.to(device)
        images = images.view(images.shape[0], -1)
        ps = model(images)
        test_loss += criterion(ps, labels)
        top_p, top_class = ps.topk(1, dim = 1)
        equals = top_class == labels.view(*top_class.shape)
        accuracy += torch.mean(equals.type(torch.FloatTensor))
    
    model.train()

    print("Epoch: {}/{}..".format(e+1, epochs),
          "Training loss: {:.3f}..".format(running_loss/len(trainloader)),
          "Test loss: {:.3f}..".format(test_loss/len(testloader)),
          "Test Accuracy: {:.3f}".format(accuracy/len(testloader)))

time2 = time.time()
print(str(time2 - time1))

/usr/local/lib/python3.7/dist-packages/torch/utils/data/dataloader.py:481: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  cpuset_checked))


19.296220779418945
Epoch: 1/5.. Training loss: 1.355.. Test loss: 0.901.. Test Accuracy: 0.726
19.082611083984375
Epoch: 2/5.. Training loss: 0.774.. Test loss: 0.716.. Test Accuracy: 0.756
19.03814721107483
Epoch: 3/5.. Training loss: 0.661.. Test loss: 0.647.. Test Accuracy: 0.772
19.375463485717773
Epoch: 4/5.. Training loss: 0.607.. Test loss: 0.608.. Test Accuracy: 0.780
19.402095079421997
Epoch: 5/5.. Training loss: 0.572.. Test loss: 0.580.. Test Accuracy: 0.795
112.22894525527954


## Overall Result

In order to run the code in appropriate manner, I have started with Script as it is with including the running time. From the result, I have obtained the following
- Epoch: 1/5.
  - Training loss: 0.488
  - Test loss: 0.423
  - Test Accuracy: 0.840
  - Time: 19.77926993370056
- Epoch: 2/5.
  - Training loss: 0.373
  - Test loss: 0.396
  - Test Accuracy: 0.854
  - Time: 19.79761290550232
- Epoch: 3/5.
  - Training loss: 0.339
  - Test loss: 0.373
  - Test Accuracy: 0.866
  - Time: 20.1186466217041
- Epoch: 4/5.
  - Training loss: 0.316
  - Test loss: 0.361
  - Test Accuracy: 0.872
  - Time: 20.131465196609497
- Epoch: 5/5.
  - Training loss: 0.297
  - Test loss: 0.371
  - Test Accuracy: 0.867
  - Time: 20.082324266433716
- Total time: 118.79080986976624

For the Batch size, I have changed following items:
- trainloader = torch.utils.data.DataLoader(trainset, batch_size = 32, shuffle = True, num_workers=4)
- testloader = torch.utils.data.DataLoader(testset, batch_size = 32, shuffle = True, num_workers=4)
- From the two codes, I have changed the batch_size from 32 to 16 and 64.

With the following changes, I have obtained the following result for batch size of 16 to be as follows: 
- Epoch: 1/5.
  - Training loss: 0.476
  - Test loss: 0.448
  - Test Accuracy: 0.837
  - Time: 27.752448081970215
- Epoch: 2/5.
  - Training loss: 0.375
  - Test loss: 0.404
  - Test Accuracy: 0.855
  - Time: 28.701561212539673
- Epoch: 3/5.
  - Training loss: 0.342
  - Test loss: 0.390
  - Test Accuracy: 0.860
  - Time: 29.306257724761963
- Epoch: 4/5.
  - Training loss: 0.317
  - Test loss: 0.367
  - Test Accuracy: 0.868
  - Time: 28.63768243789673
- Epoch: 5/5.
  - Training loss: 0.303
  - Test loss: 0.361
  - Test Accuracy: 0.874
  - Time: 28.957837343215942
- Total time: 163.80136346817017

For the case of batch size of 64, I have obtained:
- Epoch: 1/5.
  - Training loss: 0.500
  - Test loss: 0.440
  - Test Accuracy: 0.843
  - Time: 16.528880834579468
- Epoch: 2/5.
  - Training loss: 0.382
  - Test loss: 0.427
  - Test Accuracy: 0.849
  - Time: 16.736101388931274
- Epoch: 3/5.
  - Training loss: 0.347
  - Test loss: 0.371
  - Test Accuracy: 0.863
  - Time: 16.673275470733643
- Epoch: 4/5.
  - Training loss: 0.319
  - Test loss: 0.355
  - Test Accuracy: 0.873
  - Time: 16.656588792800903
- Epoch: 5/5.
  - Training loss: 0.300
  - Test loss: 0.363
  - Test Accuracy: 0.867
  - Time: 16.606858253479004
- Total time: 97.05978083610535

As one can see from the above result, as the batch size increases, the time to train and run the entire script decreases. However. it appears as if changning the batch size does not seems to affect the test accuracy. Considering training loss and test loss, there appears to be no impact.

For the Learning Rate, the change occurred in 
- optimizer = optim.Adam(model.parameters(), lr = 0.001)
- From above, I have changed lr to be 0.01 and 0.0001

With the following changes, I have obtained the following result for lr = 0.01 to be as follows: 
- Epoch: 1/5.
  - Training loss: 0.583
  - Test loss: 0.535
  - Test Accuracy: 0.817
  - Time: 20.39034152030945
- Epoch: 2/5.
  - Training loss: 0.503
  - Test loss: 0.512
  - Test Accuracy: 0.826
  - Time: 20.72736096382141
- Epoch: 3/5.
  - Training loss: 0.486
  - Test loss: 0.509
  - Test Accuracy: 0.835
  - Time: 20.818732261657715
- Epoch: 4/5.
  - Training loss: 0.473
  - Test loss: 0.548
  - Test Accuracy: 0.823
  - Time: 20.53863549232483
- Epoch: 5/5.
  - Training loss: 0.456
  - Test loss: 0.510
  - Test Accuracy: 0.834
  - Time: 20.832820177078247
- Total time: 119.51724076271057

For the case of lr = 0.0001, I have obtained:
- Epoch: 1/5.
  - Training loss: 0.651
  - Test loss: 0.514
  - Test Accuracy: 0.815
  - Time: 20.479434728622437
- Epoch: 2/5.
  - Training loss: 0.455
  - Test loss: 0.460
  - Test Accuracy: 0.835
  - Time: 20.618394374847412
- Epoch: 3/5.
  - Training loss: 0.416
  - Test loss: 0.441
  - Test Accuracy: 0.841
  - Time: 20.535950422286987
- Epoch: 4/5.
  - Training loss: 0.392
  - Test loss: 0.419
  - Test Accuracy: 0.848
  - Time: 20.329606533050537
- Epoch: 5/5.
  - Training loss: 0.374
  - Test loss: 0.409
  - Test Accuracy: 0.853
  - Time: 20.27568006515503
- Total time: 118.01114583015442

For the LR, it has quite unusual pattern. It appears that lr = 0.001 performs the best among 3, then 0.0001, and lastly 0.01. It appears for both changing in lr has impacted training loss and test loss where both losses increased in 0.01 and 0.0001 compared to lr = 0.001. In terms of time, the time to run was about the same for all three cases.


For the last case of optimizer, the change occurred in 
- optimizer = optim.Adam(model.parameters(), lr = 0.001)
- From above, Adam changed to SGD.

With the change, I have obtained the result to be
- Epoch: 1/5.
  - Training loss: 1.355
  - Test loss: 0.901
  - Test Accuracy: 0.726
  - Time: 19.296220779418945
- Epoch: 2/5.
  - Training loss: 0.774
  - Test loss: 0.716
  - Test Accuracy: 0.756
  - Time: 19.082611083984375
- Epoch: 3/5.
  - Training loss: 0.661
  - Test loss: 0.647
  - Test Accuracy: 0.772
  - Time: 19.03814721107483
- Epoch: 4/5.
  - Training loss: 0.607
  - Test loss: 0.608
  - Test Accuracy: 0.780
  - Time: 19.375463485717773
- Epoch: 5/5.
  - Training loss: 0.572
  - Test loss: 0.580
  - Test Accuracy: 0.795
  - Time: 19.402095079421997
- Total time: 112.22894525527954

Changing from Adam to SGD has decreased the run time a little bit. However, both training and test loss increased with change to SGD, and test accuracy has decreased. 

